Phoenix: Exceptions lassen Crawler nicht beenden

Wenn im Phönix-Crawler Exceptions auftreten, führt das dazu, dass der ForkJoinPool nicht sauber beendet wird.

Auszug aus Log:

> .  PHÖNIX shutdown pool...
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
com.google.gson.JsonSyntaxException: com.google.gson.JsonSyntaxException: com.google.gson.JsonSyntaxException: com.google.gson.JsonSyntaxException: com.google.gson.JsonSyntaxException: com.google.gson.stream.MalformedJsonException: Unexpected value at line 17 column 15 path $.absaetze[0].basename
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:593)
	at java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:677)
	at java.util.concurrent.ForkJoinTask.join(ForkJoinTask.java:720)
	at java.util.concurrent.ForkJoinPool.invoke(ForkJoinPool.java:2616)
	at mServer.crawler.sender.phoenix.PhoenixCrawler.addToList(PhoenixCrawler.java:45)
	at mServer.crawler.sender.MediathekReader.run(MediathekReader.java:194)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.google.gson.JsonSyntaxException: com.google.gson.stream.MalformedJsonException: Unexpected value at line 17 column 15 path $.absaetze[0].basename
	at com.google.gson.internal.Streams.parse(Streams.java:60)
	at com.google.gson.internal.bind.TreeTypeAdapter.read(TreeTypeAdapter.java:65)
	at com.google.gson.Gson.fromJson(Gson.java:887)
	at com.google.gson.Gson.fromJson(Gson.java:852)
	at com.google.gson.Gson.fromJson(Gson.java:801)
	at mServer.crawler.sender.phoenix.tasks.ZdfTaskBase.deserializeOptional(ZdfTaskBase.java:40)
	at mServer.crawler.sender.phoenix.tasks.PhoenixFilmDetailTask.processRestTarget(PhoenixFilmDetailTask.java:66)


Prüfen, ob das Problem auch in Develop auftritt und bei Bedarf Übernahme in Develop.

--- a/src/main/java/mServer/crawler/FilmeSuchen.java
+++ b/src/main/java/mServer/crawler/FilmeSuchen.java
@@ -54,6 +54,7 @@ import mServer.crawler.sender.arte.MediathekArte_de;
 import mServer.crawler.sender.arte.MediathekArte_fr;
 import mServer.crawler.sender.br.BrCrawler;
 import mServer.crawler.sender.orf.OrfCrawler;
+import mServer.crawler.sender.phoenix.PhoenixCrawler;
 
 /**
  * ###########################################################################################################
@@ -100,7 +101,7 @@ public class FilmeSuchen {
     mediathekListe.add(new MediathekSrf(this, 1));
     mediathekListe.add(new MediathekSrfPod(this, 0));
     mediathekListe.add(new OrfCrawler(this, 1));
-    //mediathekListe.add(new PhoenixCrawler(this, 1));
+    mediathekListe.add(new PhoenixCrawler(this, 1));
   }
 
   public static String[] getNamenSender() {

--- a/src/main/java/mServer/crawler/sender/phoenix/tasks/PhoenixFilmDetailTask.java
+++ b/src/main/java/mServer/crawler/sender/phoenix/tasks/PhoenixFilmDetailTask.java
@@ -63,37 +63,40 @@ public class PhoenixFilmDetailTask extends ZdfTaskBase<DatenFilm, CrawlerUrlDTO>
 
   @Override
   protected void processRestTarget(CrawlerUrlDTO aDTO, WebTarget aTarget) {
-    Optional<PhoenixFilmDetailDto> filmDetailDtoOptional = deserializeOptional(aTarget, OPTIONAL_FILM_DETAIL_DTO_TYPE_TOKEN);
-    if (!filmDetailDtoOptional.isPresent()) {
-      // tritt auf, wenn kein Film vorhanden
-      return;
-    }
-
-    PhoenixFilmDetailDto filmDetailDto = filmDetailDtoOptional.get();
-    Optional<PhoenixFilmXmlHandler> filmXmlDtoOptional = loadFilmXml(filmDetailDto.getBaseName());
-    if (!filmXmlDtoOptional.isPresent()) {
-      LOG.info("PhoenixFilmDetailTask: error parsing xml " + aDTO.getUrl());
-      return;
-    }
-
-    PhoenixFilmXmlHandler filmXmlHandler = filmXmlDtoOptional.get();
-    if (filmXmlHandler.getBaseName() == null) {
-      // tritt auf, wenn kein Film vorhanden
-      return;
-    }
-
-    Optional<DownloadDto> videoDetailDtoOptional = deserializeOptional(createWebTarget(videoDetailHost + PhoenixConstants.URL_VIDEO_DETAILS_BASE + filmXmlHandler.getBaseName()),
-            OPTIONAL_DOWNLOAD_DTO_TYPE_TOKEN);
-    if (!videoDetailDtoOptional.isPresent()) {
-      LOG.info("PhoenixFilmDetailTask: error deserializing download dto " + aDTO.getUrl());
-      return;
-    }
-
     try {
+      Optional<PhoenixFilmDetailDto> filmDetailDtoOptional = deserializeOptional(aTarget, OPTIONAL_FILM_DETAIL_DTO_TYPE_TOKEN);
+      if (!filmDetailDtoOptional.isPresent()) {
+        // tritt auf, wenn kein Film vorhanden
+        return;
+      }
+
+      PhoenixFilmDetailDto filmDetailDto = filmDetailDtoOptional.get();
+      Optional<PhoenixFilmXmlHandler> filmXmlDtoOptional = loadFilmXml(filmDetailDto.getBaseName());
+      if (!filmXmlDtoOptional.isPresent()) {
+        LOG.info("PhoenixFilmDetailTask: error parsing xml " + aDTO.getUrl());
+        return;
+      }
+
+      PhoenixFilmXmlHandler filmXmlHandler = filmXmlDtoOptional.get();
+      if (filmXmlHandler.getBaseName() == null) {
+        // tritt auf, wenn kein Film vorhanden
+        return;
+      }
+
+      Optional<DownloadDto> videoDetailDtoOptional = deserializeOptional(createWebTarget(videoDetailHost + PhoenixConstants.URL_VIDEO_DETAILS_BASE + filmXmlHandler.getBaseName()),
+              OPTIONAL_DOWNLOAD_DTO_TYPE_TOKEN);
+      if (!videoDetailDtoOptional.isPresent()) {
+        LOG.info("PhoenixFilmDetailTask: error deserializing download dto " + aDTO.getUrl());
+        return;
+      }
+
       addFilm(filmDetailDto, filmXmlHandler, videoDetailDtoOptional.get());
 
     } catch (MalformedURLException e) {
       LOG.error("PhoenixFilmDetailTask: url can't be parsed: ", e);
+    } catch (Exception e) {
+      // catch all exceptions to ensure that the crawler can process the other results
+      LOG.fatal(e);
     }
   }
 

--- a/src/main/java/mServer/crawler/sender/phoenix/tasks/PhoenixOverviewTask.java
+++ b/src/main/java/mServer/crawler/sender/phoenix/tasks/PhoenixOverviewTask.java
@@ -42,17 +42,21 @@ public class PhoenixOverviewTask extends ZdfTaskBase<CrawlerUrlDTO, CrawlerUrlDT
 
   @Override
   protected void processRestTarget(CrawlerUrlDTO aDTO, WebTarget aTarget) {
-    Optional<SendungOverviewDto> overviewDtoOptional = deserializeOptional(aTarget, OPTIONAL_OVERVIEW_DTO_TYPE_TOKEN);
-    if (!overviewDtoOptional.isPresent()) {
-      LOG.fatal("PhoenixOverviewTask: error processing url " + aDTO.getUrl());
-      return;
-    }
+    try {
+      Optional<SendungOverviewDto> overviewDtoOptional = deserializeOptional(aTarget, OPTIONAL_OVERVIEW_DTO_TYPE_TOKEN);
+      if (!overviewDtoOptional.isPresent()) {
+        LOG.fatal("PhoenixOverviewTask: error processing url " + aDTO.getUrl());
+        return;
+      }
 
-    SendungOverviewDto overviewDto = overviewDtoOptional.get();
-    addResults(overviewDto.getUrls());
+      SendungOverviewDto overviewDto = overviewDtoOptional.get();
+      addResults(overviewDto.getUrls());
 
-    if (overviewDto.getNextPageId().isPresent()) {
-      taskResults.addAll(createNewOwnInstance(baseUrl + overviewDto.getNextPageId().get()).invoke());
+      if (overviewDto.getNextPageId().isPresent()) {
+        taskResults.addAll(createNewOwnInstance(baseUrl + overviewDto.getNextPageId().get()).invoke());
+      }
+    } catch (Exception e) {
+      LOG.fatal(e);
     }
   }
 

